### ISSUE TITLE
Shortest path algorithms: unintentional use of a negative cycle detection heuristic

### ISSUE THREAD

**Issue by kazimierz-256** [networkx/networkx#4047]
Relevant pull request: https://github.com/networkx/networkx/pull/3879

@dschult I've just noticed that the newly contributed negative cycle detection heuristic is unintentionally used by default in every method that uses the helper method `_bellman_ford`, namely:

- johnson
- bellman_ford_path_length
- single_source_bellman_ford
- single_source_bellman_ford_path_length
- bellman_ford_predecessor_and_distance

Only in the latter case the user can opt in with the heuristic, otherwise it is used by default. This is because the helper method `_bellman_ford` has a new default parameter `heuristic` set to `True` and the first 4 mentioned methods do not set this new default parameter to any value.

We can fix this in 2 ways:
1. By adding the ability to opt in with the heuristic to every method that internally uses `_bellman_ford`, or
2. By using the heuristic always by default.

I'm an advocate of the 2nd solution for two reasons:
- Performance: The most expensive operations are dict lookups. If the graph does not contain a negative cycle, the empirical difference in performance is just a few percent. If the graph indeed contains a negative cycle, the speedup is 10X, 20X and in some cases even 30X.
- Simplicity: The end user is no longer baffled whether they should opt in with the heuristic or not.

**Comment by dschult** [comment]
Yes, After looking through this again, there are other functions (that you mention) that call this code and didn't get updated to turn off the new option. 

Also, the part of the code turned on by ```heuristic``` is not actually a heuristic at all (it doesn't approximate the solution in any way, it just stores some extra info that helps to identify negative cycles).  So really, the name of that keyword should be more specific to this case rather than "heuristic".

Standard Bellman-Ford doesn't use this cycle finding short-cut, so I think we should allow it to be turned off. The cost of storing the info and checking for short cycles depends highly on the structure of the graph and could be much more for some graph types. Especially considering that many use cases know ahead of time that there aren't any negative cycles we want to allow people to turn this off and just use a plain, ordinary Bellman-Ford. Of course, some cases, like ```negative_edge_cycle``` could simply turn it on always... 

What's a better name for the keyword flag than "heuristic"?    It's a shortcut that relies on storing/caching the short cycles.
"short_cycle_cache"?   "negative_cycle_shortcut"?   "cycle_cache"?    "cache"?

**Comment by kazimierz-256** [comment]
Hi, sorry for the delay. Now I finally get it: the notion of a heuristic in the literature actually relates to an optimized way of extracting the next vertex from the queue by the Bellman-Ford algorithm.

In my opinion the most accurate name is a brief description of what the idea does from the point of the end user:
- "negative_cycle_early_detection_heuristic" or
- "negative_cycle_detection_heuristic" or
- "negative_cycle_early_detection"

Which of the above keywords would you be most comfortable with?

**Comment by dschult** [comment]
Hi.. thanks for these ideas --  I like "negative_cycle_early_detection" the best. But it is long.... :}

I'm also trying to understand the shortcut process better. In particular I want to make sure that I didn't break it inadvertently when I re-factored your code. My understanding is based on the comments you put in the code that I then changed also, so I'm trying to reconstruct how you intended this to work before all my changes  :{

Here's my understanding:  we use 2 data structures to look for shortcuts:  
- ```pred_edge``` is a dict keyed by node "v" to the edge coming into "v" along the current shortest path to v. I changed the value to be the predecessor of "v" on that edge because we never used the entry "v" in that edge.  So, the merged code actually has this as a dict keyed by node "v" to the predecessor of "v" on the current shortest path to v.  I think this might be the same as ```pred``` only ```pred``` handles ties better.
- ```recent_update``` is a dict keyed by node "v" to an edge on the current shortest path from a source node to "v". Thus if the source nodes are 0, 1 and there is a current shortest path [1, 3, 5] then ```recent_update[5]``` could return (1,3) or (3,5).  It will return (3, 5) unless that edge has been used to get a path to "5" already. If this is a second path that uses edge (3,5) then recent_update[5] returns (1,3).
Is this correct?

Then the logic of the shortcut is: 
if we already came to node v along the edge from u, then set ```recent_update[v]``` to be ```recent_update[u]```.

Can you explain how this shortcut works?  I think I understood it before, but am not so sure anymore. :}
Thanks!

**Comment by kazimierz-256** [comment]
Hi, please give me 2 more weeks to reply to this thread. I'm simultaneously
working full-time, studying and moving to a new city. In september I'll be
more available. Thanks for your understanding, we'll fix this eventually!

**Comment by rossbar** [comment]
Gentle ping @kazimierz-256 . Is this issue still valid? There have been some developments on the negative cycle detection front that may be relevant here: see #5089

**Comment by kazimierz-256** [comment]
> Is this issue still valid?

Yes, now I see it concerns:
- johnson
- bellman_ford_path_length
- single_source_bellman_ford
- single_source_bellman_ford_path
- single_source_bellman_ford_path_length
- bellman_ford_predecessor_and_distance
- bellman_ford_path

> I think this might be the same as pred only pred handles ties better.

Great idea! To handle ties, now, `pred_edge[v]` ought to be a set of vertices, for which all edges `(u,v) for u in pred_edge[v]` lie on shortest paths leading to "v".

Now, upon [discovering another shortest path](https://github.com/networkx/networkx/blob/d8e76beada5e3a962ef95c4b549746e493bb662f/networkx/algorithms/shortest_paths/weighted.py#L1449) one ought to union the set `recent_update[u]` with (now, also a set) `recent_update[v]`.
Now, [negative cycle detection](https://github.com/networkx/networkx/blob/d8e76beada5e3a962ef95c4b549746e493bb662f/networkx/algorithms/shortest_paths/weighted.py#L1422) would no longer be a constant time operation, but O(V) in the worst case: if there ever occurred a cataclysmic amount of equivalent shortest paths 🤔.

> I like "negative_cycle_early_detection" the best. But it is long.... :}

How about:
- "negative_cycle_predetection",
- "negative_cycle_predetection_cache",
- "negative_cycle_cache",
- "seek_negative_cycles",
- "track_negative_cycles",
- "trace_negative_cycles" 🙂?,
- "pretrace_negative_cycles",
- "trail_negative_cycles",
- "negative_cycle_tracker",
- "negative_cycle_tracer",
- "negative_cycle_trailer".

> If this is a second path that uses edge (3,5) then recent_update[5] returns (1,3).
Is this correct?

Yes 🙂, in that case no negative cycle would be detected. It could also return (5,1)... and this is where we find the v="5" in the edge in `recent_update[5]` and this is the very moment the existence of a negative cycle is detected.

> if we already came to node v along the edge from u, then set recent_update[v] to be recent_update[u].

Yes. The main idea is to keep track of the earliest occurrence of a repeated edge relaxation in the hope of encountering one of the vertices of this cached edge (which may happen only along a doubly-relaxed-path). And such a self-connecting doubly-relaxed-path is, simply put, the cycle we were looking for.
Cool, right? 🙂

> The cost of storing the info and checking for short cycles depends highly on the structure of the graph and could be much more for some graph types.

Why do you think so? Which graphs would induce computational trouble?
If `pred_edge[v]` is just a node, and `recent_update[v]` is just an edge, then the [average-case complexity is constant](https://wiki.python.org/moin/TimeComplexity#:~:text=k%20in%20d-,O(1),-O(n)) for all appended instructions. Since e.g. the instruction [`dist[v] = dist_v`](https://github.com/networkx/networkx/blob/d8e76beada5e3a962ef95c4b549746e493bb662f/networkx/algorithms/shortest_paths/weighted.py#L1445) is always executed, no computational complexity deterioration should occur on average.
Still, if this is a concern, I'll refactor the code to disable the default use of the heuristic. Or perhaps just leave an option for the user to opt-out if they operate on some extremely unusual graphs (if they exist)?
